pallets/dispenser/src/lib.rs

```diff
@@ -0,0 +1,187 @@
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		/// The loose coupling to a vesting schedule implementation.
+		type VestingSchedule: VestingSchedule<Self::AccountId, Moment = BlockNumberFor<Self>>;
+		/// The period of time that the dispensed funds are in a vesting schedule. The schedule
+		/// starts after the lock period.
+		#[pallet::constant]
+		type VestPeriod: Get<BlockNumberFor<Self>>;
+		/// The Ed25519 Verifier Public Key to verify the signature of the credentials.
+		#[pallet::constant]
+		type VerifierPublicKey: Get<[u8; 32]>;
+		/// A type representing the weights required by the dispatchables of this pallet.
+		type WeightInfo: crate::weights::WeightInfo;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	pub type DispenseAmount<T> = StorageValue<_, BalanceOf<T>, ValueQuery, <T as Config>::InitialDispenseAmount>;
+
+	#[pallet::storage]
+	pub type Dispensed<T> = StorageMap<_, Blake2_128Concat, Did, ()>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		Dispensed { dispensed_to_did: Did, dispensed_to: T::AccountId, amount: BalanceOf<T> },
+		DispenseAmountChanged(BalanceOf<T>),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The dispenser has already dispensed to the did.
+		DispensedAlreadyToDid,
+		/// The dispenser account does not have any funds to distribute.
+		DispenserDepleted,
+		/// The dispense amount is too low. It must be greater than the free dispense amount.
+		DispenseAmountTooLow,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::feeless_if( | origin: &OriginFor<T>, jwt: &UntrustedToken | -> bool {
+            if let Ok((_, did, _)) = T::InvestorOrigin::ensure_origin(origin.clone(), jwt, T::VerifierPublicKey::get()) {
+                return Dispensed::<T>::get(did).is_none()
+            } else {
+                return false
+            }
+        })]
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::dispense())]
+		pub fn dispense(origin: OriginFor<T>, jwt: UntrustedToken) -> DispatchResultWithPostInfo {
+			let (who, did, _investor_type) =
+				T::InvestorOrigin::ensure_origin(origin, &jwt, T::VerifierPublicKey::get())?;
+			ensure!(Dispensed::<T>::get(&did).is_none(), Error::<T>::DispensedAlreadyToDid);
+
+			let amount = DispenseAmount::<T>::get();
+			ensure!(CurrencyOf::<T>::free_balance(&Self::dispense_account()) >= amount, Error::<T>::DispenserDepleted);
+
+			let current_block = <frame_system::Pallet<T>>::block_number();
+			let length_as_balance = T::BlockNumberToBalance::convert(T::VestPeriod::get());
+			let locked_amount = amount.saturating_sub(T::FreeDispenseAmount::get());
+			let per_block = locked_amount / length_as_balance.max(sp_runtime::traits::One::one());
```

Comment:
Is it safe to use the `/` here? Or should we use safe/checked math? What happens when `locked_amount / length_as_balance` is a float?

Reply:
I think this is safe to use as I copied the logic from a parity pallet, but I can change it to safe math just to be sure!